lib/streamlit/runtime/runtime.py

```diff
@@ -335,7 +346,7 @@ def is_active_session(self, session_id: str) -> bool:
     def connect_session(
         self,
         client: SessionClient,
-        user_info: dict[str, str | None],
+        user_info: dict[str, str | bool | None],
```

Comment:
Can you help me understand the need for the bool?

Reply:
Sure!

As PM rightly noted in the spec, potentially, email could not be included in the Identity provider response. This is very unlikely to happen because we always by default request `email` scope, and `email` is part of OIDC standard claims https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

However, to protect ourselves from identity providers that don't include email, I decided not to rely on the `email` field to verify successful authentication. Instead, in case of a successful authentication response, we add the `_streamlit_logged_in` field into the `user_info` of `AppSession` (we also set that field to True in case of not empty/placeholder email from the Cloud header). 

Then we compute the `is_authenticated` property method of `experimental_user` based on that explicit value.

To not raise any typing errors, we describe explicitly here that `user_info` could also carry a `bool` value. 
These changes are for mypy, on the runtime time they don't affect what exact values we store in `user_info`.